### PR TITLE
SMPP transport handles throttling by message instead of PDU

### PR DESCRIPTION
**Pull request by jerith**
This is a problem when we have multiple PDUs per message.

### CODE REVIEW

File: vumi/transports/smpp/smpp_service.py [hunk 1]
```diff
@@ -60,28 +71,152 @@ def stopService(self):
...
+                "Could not retrieve throttled pdu: %s" % (seq_no,))
+            self.check_stop_throttling(0)
+        else:
+            # Try handle this message again and leave the rest to our
+            # submit_sm_resp handlers.
+            log.msg("Retrying throttled pdu for message: %s" % (
+                pdu_data.vumi_message_id,))
+            # This is a new PDU, so it needs a new sequence number.
+            new_seq_no = yield self.sequence_generator.next()
+            pdu_data.pdu.obj['header']['sequence_number'] = new_seq_no
```
Comment (smn) [inline vumi/transports/smpp/smpp_service.py]: `python-smpp` :(

File: vumi/transports/smpp/tests/test_smpp_transport.py [hunk 2]
```diff
@@ -921,6 +969,45 @@ def test_mt_sms_tps_limits(self):
         self.assertEqual(short_message(submit_sm_pdu3), 'hello world 3')
 
     @inlineCallbacks
+    def test_mt_sms_tps_limits_multipart(self):
+        """
+        TPS throttling counts PDUs, but doesn't finishes sending the current
+        message.
```
Comment (smn) [inline vumi/transports/smpp/tests/test_smpp_transport.py]: `s/finishes/finish/`

File: vumi/transports/smpp/tests/test_smpp_transport.py [hunk 3]
```diff
@@ -921,6 +969,45 @@ def test_mt_sms_tps_limits(self):
...
+        with LogCatcher(message="No longer throttling outbound") as lc:
+            self.clock.advance(1)
+        [logmsg] = lc.logs
+        self.assertEqual(logmsg['logLevel'], logging.INFO)
+        self.assertEqual(transport.throttled, False)
+
+        yield msg3_d
+        [pdu3_1, pdu3_2] = yield self.fake_smsc.await_pdus(2)
+        self.assertEqual(short_message(pdu3_1)[-5:], '33333')
+        self.assertEqual(short_message(pdu3_2)[-5:], '3333c')
```
Comment (smn) [inline vumi/transports/smpp/tests/test_smpp_transport.py]: Clever way of testing this.

**Comment by smn** [comment]
Considering `_throttled_pdus` is an in-memory list, what will happen if we restart while we're being throttled? If we're doing 100 msg/s will we potentially be dropping `100 * n` messages? Where `n` is the amount of seconds we've been throttled at that point.

File: vumi/transports/smpp/smpp_service.py (hunk 1 above)
Comment (jerith) [inline vumi/transports/smpp/smpp_service.py]: Yeah. Reimplementing that wasn't a yak I needed to shave this time.

File: vumi/transports/smpp/tests/test_smpp_transport.py (hunk 2 above)
Comment (jerith) [inline vumi/transports/smpp/tests/test_smpp_transport.py]: Thanks.

File: vumi/transports/smpp/tests/test_smpp_transport.py (hunk 3 above)
Comment (jerith) [inline vumi/transports/smpp/tests/test_smpp_transport.py]: :-)

**Comment by jerith** [comment]
> Considering `_throttled_pdus` is an in-memory list, what will happen if we restart while we're being throttled? If we're doing 100 msg/s will we potentially be dropping `100 * n` messages? Where `n` is the amount of seconds we've been throttled at that point.

Correct. This is the existing behaviour, except with PDUs instead of messages. Changing that is out of scope for this PR.

File: vumi/transports/smpp/smpp_transport.py [hunk 4]
```diff
@@ -178,6 +198,20 @@ def get_cached_message(self, message_id):
     def delete_cached_message(self, message_id):
         return self.redis.delete(message_key(message_id))
 
+    def cache_pdu(self, vumi_message_id, seq_no, pdu):
+        # We the PDU in wire format to avoid json encoding troubles.
+        key = pdu_key(seq_no)
+        expiry = self.config.submit_sm_expiry
+        return self.redis.setex(key, expiry, json.dumps(
+            {'vumi_message_id': vumi_message_id, 'pdu': pdu.get_hex()}))
```
Comment (hodgestar) [inline vumi/transports/smpp/smpp_transport.py]: Should we move this `json.dumps(...)` into a `.to_json(..)` method on `CachedPDU`?

Comment (jerith) [inline vumi/transports/smpp/smpp_transport.py]: That would be cleaner, yes. Thanks!

**Comment by hodgestar** [comment]
:+1:

**Pull request merged**